### DISCUSSION TITLE
CoolProp now available in pyodide and juypterlite

### DISCUSSION THREAD

**Discussion by dvd101x** [CoolProp/CoolProp#2302] [General]
Hi, just to let you know that now it is possible to use CoolProp from JupyterLite (jupyter from the web without installing anything) thanks to the CoolProp addition to the pyodide project which I think is really cool and opens a lot of possibilities.

https://jupyterlite.readthedocs.io

**Comment by ibell** [comment]
Whoah this is awesome. How was this done and who did it? I would like to replicate this for some other libraries I am developing.

**Reply by dvd101x** [reply]
It all comes from here https://github.com/pyodide/pyodide/pull/4028

made by https://github.com/nascheinkman

Previously it wasn’t possible due to CoolProp not having pure python wheels. With this PR CoolProp worked on pyodide, which was later updated in pyodide kernel and then updated in jupyter lite.

**Reply by ibell** [reply]
Can someone please add some docs to CoolProp about this?

**Reply by dvd101x** [reply]
I could do it. Do you have something in mind? Like an example in particular, instructions on how to run it, etc. ?

Or just like include jupyterlite in the list of websites that uses CoolProp

As pyodide includes numpy, matplotlib and [others](https://pyodide.org/en/stable/usage/packages-in-pyodide.html#packages-in-pyodide) an example could be made with a thermodynamic cycle of your choosing or maybe even one that is dynamic similar to https://observablehq.com/@dvd101x/vapor-compression-cycle (WIP)

One possibility is also for CoolProp to host it's own jupyter lite so that includes it's notebooks.

@ibell just for curiosity, what other libraries do you have in mind?

**Reply by ibell** [reply]
Experimenting a bit I am dismayed to see that exceptions kill the entire instance. That's not good. I have had similar experiences with emscripten compiled code to wasm, and there are flags that allow for handling of exceptions.

I would like to add ChebTools (for playing with Chebyshev expansions), and teqp (yet another EOS library) and perhaps a few others

**Reply by dvd101x** [reply]
Ohh, I did some tests and for simple stuff it was working, but for more complex stuff it didn't, I didn't notice why.

Maybe some fix could be implemented directly on pyodide.

Also it is possible to import the javascript implementation of coolprop to pyodide if needed.

```html
<!doctype html>
<html>

<head>
    <script src="https://cdn.jsdelivr.net/npm/coolporp@1.0.0/coolprop.js"></script>
    <script src="https://cdn.jsdelivr.net/pyodide/dev/full/pyodide.js"></script>
</head>

<body>
    Pyodide test page <br>
    Open your browser console to see Pyodide output
    <script type="text/javascript">
        async function main() {
            let pyodide = await loadPyodide();
            console.log(pyodide.runPython(`
            import sys
            sys.version
            `));
            pyodide.runPython(`
            import js
            PropsSI = js.Module.PropsSI
            HAPropsSI = js.Module.HAPropsSI

            phases = ["liquid", "supercritical", "supercritical_gas", "supercritical_liquid", "critical_point", "gas", "twophase", "unknown", "not_imposed"]

            def PhaseSI(prop1, val1, prop2, val2, fluid):
                return phases[PropsSI("Phase", prop1, val1, prop2, val2, fluid)]

            # Density of Nitrogen at 298.15 K and 101325 Pa in kg/m^3
            print(PropsSI('D', 'T', 298.15, 'P', 101325, 'Nitrogen'))

            # Enthalpy (J per kg dry air) as a function of temperature, pressure,
            #    and relative humidity at dry bulb temperature T of 25C, pressure
            #    P of one atmosphere, relative humidity R of 50%
            print(HAPropsSI("H", "T", 298.15, "P", 101325, "R", 0.5))

            # Phase of water at 101325 Pa and Quality 0%
            print(PhaseSI("P", 101325, "Q", 0, "Water"))
            `)
        }
        main();
    </script>
</body>

</html>
```

**Reply by ibell** [reply]
I think in the build, exception-handling flags need to be built. Can't recommend yet until that gets resolved.  For teqp, this is the chunk that I needed in the CMake file: https://github.com/usnistgov/teqp/blob/53392a4d659f95e3eb0d264db60f7777e97bde80/CMakeLists.txt#L259

**Reply by dvd101x** [reply]
Hi,

I'm happy to report some progress on the [pyodide PR](https://github.com/pyodide/pyodide/pull/4455). I incorporated the exception-handling flags you suggested, and it seems to be a positive step forward.

There are still some tests failing in Chrome and Node. It's possible these might be resolved once they update their continuous integration (CI) system.  While I don't have prior experience with this type of issue, the pyodide team has been helpful in guiding me through the PR process.

For now, I'm monitoring the situation and making occasional updates. I'll keep you posted on any further developments.

**Reply by dvd101x** [reply]
I'm happy to report that the [PR](https://github.com/pyodide/pyodide/pull/4455) was accepted and it's now part of the new release.

https://github.com/pyodide/pyodide/releases/tag/0.26.0

**Comment by mhvwerts** [comment]
Wow. It really works! For regular users of https://jupyterlite.readthedocs.io/en/latest/_static/lab/index.html, be sure to click the 'Reload' button while holding the shift key, so that Jupyterlite is updated to the latest version.

Afterwards, a Jupyter(lite) Notebook can be created and we can simply
```python
from CoolProp.CoolProp import PropsSI
print(PropsSI('V', 'T', 298.15, 'P', 101325., 'water'))
```
to get the viscosity of water at STP, for instance, or any other thermophysical calculation possible through the great CoolProp.

Many thanks, this is very helpful to many students, teachers, scientists

**Reply by dvd101x** [reply]
Please send thanks to https://github.com/pyodide/pyodide/pull/4028

**Comment by ibell** [comment]
Sounds promising! Thanks for the update.

**Comment by ibell** [comment]
Great work!

**Reply by dvd101x** [reply]
Thanks!

Now I see the update working in [jupyterlite ](https://jupyterlite.readthedocs.io/en/stable/) and [pyscript](https://pyscript.net/).

With this maybe some examples could be included in this project. Previously we had some discussions about it that we could continue.

**Reply by ibell** [reply]
Go for it. I cannot provide any help, but I can review pull requests.

**Reply by dvd101x** [reply]
Thanks Ian!

Will do. It might take a while.
